Message when date range is restricted and there are no results

jeanlange opened the pull request:
Suggest to the user that they should enlargen their date range
if there are no results and they have not the largest date range
selected.

Commit:
Message when date range is restricted and there are no results

Suggest to the user that they should enlargen their date range
if there are no results and they have not the largest date range
selected.

jeanlange commented:
Here's the effect of this one.

<img width="558" alt="discovery_demo_" src="https://cloud.githubusercontent.com/assets/1461735/23680042/d354f076-0357-11e7-8f69-409f7873aaee.png">

kognate commented on views/Query/Collapsed.jsx:
```diff
@@ -97,20 +100,28 @@ export default React.createClass({
-    this.onDatesChange(newDates);
+    Promise.resolve(
+      this.setState({restrictedDateRange: restrictedDateRange})
+    ).then(() => {
```
why the promise?

jeanlange replied:
Because without it, the local state didn't update in time for onDatesChange to see the updated value of restrictedDateRange, and it remained false (or true, if it was true) until you clicked on the button a second time. You could see this happen if you comment out the promise lines and watch the query state as you click around.

There's a similar bug with the dates where sometimes it doesn't update the dates on the first click because the state update doesn't happen fast enough, which I have just realized. I'll fix that in another commit :D

jeanlange replied:
another commit added.

kognate replied:
I would argue that this kind of sequence dependency is better handled by the React component lifecycle methods than a promise (like `componentDidUpdate`)

jeanlange replied:
handled by using a callback to setState, as discussed. See third commit.

kognate reviewed (changes requested):
I'm not sure that promise is needed, and I've requested clarification.

Commit:
Wait for date state to be set before relying on it

Before this commit, sometimes clicking on a date button would not
update the results because the query would run before the date state
updated - you'd have to click it again to see the change.

kognate commented on views/Query/Collapsed.jsx:
```diff
@@ -55,8 +55,7 @@ export default React.createClass({
 
-  onDatesChange({ startDate, endDate }) {
-    this.setState({ startDate, endDate });
+  onDatesChange() {
```
onDatesChange has a signature change (removing the first arg) but is called with an argument on line 126.

jeanlange replied:
fixed, missed this.

kognate reviewed (changes requested):
method signature change question

kognate commented on views/NoResults/index.jsx:
```diff
@@ -6,6 +6,9 @@ const NoResults = (props) => (
     <p className="base--h2">
       There are no results for: {props.query ? props.query.text : '' }
     </p>
+    <p className="base--h3" style={props.query.restrictedDateRange ? {display: 'block'} : {display: 'none'}}>
```
Also, this is mostly just a style comment,  I would change line 9 to be something more like:

`<p className={props.restrictedDateRange ? 'base--h3' : 'base--h3 hidden'}>`

instead of embedded `style`.

jeanlange replied:
agree, thanks.

kognate commented on views/Query/index.jsx:
```diff
@@ -14,6 +14,7 @@ Query.propTypes = {
     date: React.PropTypes.object,
   }),
   onQueryChange: React.PropTypes.func.isRequired,
+  dateRangeRestricted: React.PropTypes.bool,
```
You added this prop,  but no prop by this name shows up in the rest of the PR.

jeanlange replied:
moved this to the appropriate place, missed this one.

I put it in to show a reader that this property could get added to a query object. I /think/ that's appropriate, but am not a React expert.

Commit:
change promise to setstate callback

for more idiomatic react!

kognate reviewed (approved):
approve
